### PR TITLE
Freeze

### PR DESCRIPTION
**Pull request by crhea93**
I've redone how we freeze the broadening and velocity parameters based off a previous fit.

We need to add the ability to bin the previous fits. We also need documentation/examples.

- [ ] Binning
- [ ] Example
- [ ] Documentation
- [ ] Update `cube.fit_region()`
- [ ] Update `cube.fit_spectrum_region()`

### CODE REVIEW

**Comment by crhea93** [comment]
**BUG** 
The code needs to be updated to include the same logic for `cube.fit_spectrum_region()`. Right now it errors out on this function :)

File: LUCI/LuciFit.py [hunk 1]
```diff
@@ -326,36 +330,46 @@ def line_vals_estimate(self, line_name):
         """
         line_theo = self.line_dict[line_name]
         if self.ML_model is None or self.ML_model == '':
-            if self.initial_values is not False:
+            if self.freeze is True:
```
Comment (JustineGiroux) [inline LUCI/LuciFit.py:333]: ``if self.freeze:``

File: LUCI/LuciFit.py [hunk 2]
```diff
@@ -326,36 +330,46 @@ def line_vals_estimate(self, line_name):
...
                 self.vel_ml = self.initial_values[0]  # Velocity component of initial conditions in km/s
                 self.broad_ml = self.initial_values[1]  # Broadening component of initial conditions in km/s
             else:
                 pass
         else:
-            pass  # vel_ml and broad_ml already set using ML algorithm
```
Comment (JustineGiroux) [inline LUCI/LuciFit.py:335]: else: pass combo is unnecessary

File: LUCI/LuciFit.py [hunk 3]
```diff
@@ -326,36 +330,46 @@ def line_vals_estimate(self, line_name):
...
-        line_pos_est = 1e7 / ((self.vel_ml / SPEED_OF_LIGHT) * line_theo + line_theo)  # Estimate of position of line in cm-1
+            if self.freeze is True:
+                self.vel_ml = self.initial_values[0]  # Velocity component of initial conditions in km/s
+                self.broad_ml = self.initial_values[
+                    1]  # Broadening component of initial conditions in km/s
```
Comment (JustineGiroux) [inline LUCI/LuciFit.py:342]: why is this 2 lines?

File: LUCI/LuciFit.py [hunk 4]
```diff
@@ -326,36 +330,46 @@ def line_vals_estimate(self, line_name):
...
-            if self.initial_values is not False:
+            if self.freeze is True:
                 self.vel_ml = self.initial_values[0]  # Velocity component of initial conditions in km/s
                 self.broad_ml = self.initial_values[1]  # Broadening component of initial conditions in km/s
             else:
                 pass
         else:
-            pass  # vel_ml and broad_ml already set using ML algorithm
-        line_pos_est = 1e7 / ((self.vel_ml / SPEED_OF_LIGHT) * line_theo + line_theo)  # Estimate of position of line in cm-1
+            if self.freeze is True:
```
Comment (JustineGiroux) [inline LUCI/LuciFit.py:339]: ``if self.freeze:``

File: LUCI/LuciFit.py [hunk 5]
```diff
@@ -542,40 +558,63 @@ def calculate_params(self):
...
+                lines_fit.append(self.lines[mod])  # Add to list of lines fit
+                amp_est, vel_est, sigma_est = self.line_vals_estimate(self.lines[mod])  # Estimate initial values
+                initial[mod] = amp_est - initial[-1]  # Subtract continuum estimate from amplitude estimate
+                initial_positions[mod] = vel_est
+                initial_sigmas[mod] = sigma_est
+            soln = minimize(nll, initial,
+                            method='SLSQP',
+                            options={'disp': False, 'maxiter': 30},
+                            tol=1e-2,
+                            args=())
```
Comment (JustineGiroux) [inline LUCI/LuciFit.py:604]: the minimize could be out of the if/else clause and have constraints=[] for the else case

File: LUCI/LuciFit.py [hunk 6]
```diff
@@ -326,36 +330,46 @@ def line_vals_estimate(self, line_name):
...
         if self.mdn:
             # Update position and sigma_gauss bounds -- looks gross but it's the usual transformation
-            self.x_min = 1e7 / (((self.vel_ml + 3 * self.vel_ml_sigma) / SPEED_OF_LIGHT) * line_theo + line_theo)  # Estimate of position of line in cm-1
-            self.x_max = 1e7 / (((self.vel_ml - 3 * self.vel_ml_sigma) / SPEED_OF_LIGHT) * line_theo + line_theo)  # Estimate of position of line in cm-1
-            self.sigma_min = (line_pos_est * self.broad_ml) / SPEED_OF_LIGHT - 3 * (line_pos_est * self.broad_ml_sigma) / SPEED_OF_LIGHT
-            self.sigma_max = (line_pos_est * self.broad_ml) / SPEED_OF_LIGHT + 3 * (line_pos_est * self.broad_ml_sigma) / SPEED_OF_LIGHT
+            self.x_min = 1e7 / (((
+                                             self.vel_ml + 3 * self.vel_ml_sigma) / SPEED_OF_LIGHT) * line_theo + line_theo)  # Estimate of position of line in cm-1
+            self.x_max = 1e7 / (((
+                                             self.vel_ml - 3 * self.vel_ml_sigma) / SPEED_OF_LIGHT) * line_theo + line_theo)  # Estimate of position of line in cm-1
```
Comment (JustineGiroux) [inline LUCI/LuciFit.py:368]: there appears to be a lot of blank space here

File: LUCI/LuciFit.py [hunk 7]
```diff
@@ -542,40 +558,63 @@ def calculate_params(self):
...
-                        options={'disp': False, 'maxiter': 100},
-                        tol=1e-4,
-                        args=(), constraints=cons
-                        )
         parameters = soln.x
         # We now must unscale the amplitude
-        for i in range(self.line_num):
-            parameters[i * 3] *= self.spectrum_scale
-            self.uncertainties[i * 3] *= self.spectrum_scale
+        if self.freeze is True:  # Freezing velocity and broadening
```
Comment (JustineGiroux) [inline LUCI/LuciFit.py:610]: ``if self.freeze:``

File: LUCI/LuciFit.py [hunk 8]
```diff
@@ -542,40 +558,63 @@ def calculate_params(self):
...
+            for i in range(self.line_num):
+                parameters[i] *= self.spectrum_scale
+                self.uncertainties[i] *= self.spectrum_scale
+        else:  # Not freezing velocity and broadening
+            for i in range(self.line_num):
+                parameters[i * 3] *= self.spectrum_scale
+                self.uncertainties[i * 3] *= self.spectrum_scale
```
Comment (JustineGiroux) [inline LUCI/LuciFit.py:617]: to be cleaner you could merge the 2 for loops and the if could be used to determine the index set

File: LUCI/LuciFit.py [hunk 9]
```diff
@@ -589,96 +628,31 @@ def calculate_params(self):
             except np.linalg.LinAlgError:
                 covariance_mat = -np.linalg.pinv(hessian_calc)
                 self.uncertainties = np.sqrt(np.abs(np.diagonal(covariance_mat)))
-
-        self.fit_sol = parameters
+        if self.freeze is False:  # Not freezing velocity and broadening so nothing needs to be done
```
Comment (JustineGiroux) [inline LUCI/LuciFit.py:631]: ``if self.freeze:``

File: LuciBase.py [hunk 10]
```diff
@@ -236,11 +237,88 @@ def fit_entire_cube(self, lines, fit_function, vel_rel, sigma_rel, bkg=None, bin
         y_max = self.cube_final.shape[1]
         self.fit_cube(lines, fit_function, vel_rel, sigma_rel, x_min, x_max, y_min, y_max)
 
+    def fit_calc(self, i, x_min, x_max, y_min, fit_function, lines, vel_rel, sigma_rel, bayes_bool=False,
+                 bayes_method='emcee',
+                 uncertainty_bool=False, nii_cons=False,
+                 bkg=None, binning=None, spec_min=None, spec_max=None, initial_values=False):
```
Comment (JustineGiroux) [inline LuciBase.py:243]: missing doc?

File: LuciBase.py [hunk 11]
```diff
@@ -236,11 +237,88 @@ def fit_entire_cube(self, lines, fit_function, vel_rel, sigma_rel, bkg=None, bin
...
+            sky = self.cube_final[x_pix, y_pix, :]
+            if bkg is not None:  # If there is a background variable subtract the bkg spectrum
+                if binning:  # If binning, then we have to take into account how many pixels are in each bin
+                    sky -= bkg * binning ** 2  # Subtract background spectrum
+                else:  # No binning so just subtract the background directly
+                    sky -= bkg  # Subtract background spectrum
+            good_sky_inds = [~np.isnan(sky)]  # Find all NaNs in sky spectrum
+            sky = sky[good_sky_inds]  # Clean up spectrum by dropping any Nan values
+            axis = self.spectrum_axis#[good_sky_inds]  # Clean up axis  accordingly
+            if initial_values[0] is not False:   #Frozen parameter
```
Comment (JustineGiroux) [inline LuciBase.py:272]: ``if not initial_values[0]:``

**Review by JustineGiroux — COMMENTED** [review]
@crhea93 Great job!!!

File: LUCI/LuciFit.py (hunk 5 above)
Comment (crhea93) [reply LUCI/LuciFit.py:604]: I tried that but constraints being empty was throwing errors.

File: LUCI/LuciFit.py (hunk 8 above)
Comment (crhea93) [reply LUCI/LuciFit.py:617]: excellent idea!

**Pull request merged**